From 86617950a93a3cf8cecade8f648786ab40d0177a Mon Sep 17 00:00:00 2001
From: John Johansen <john.johansen@canonical.com>
Date: Wed, 10 Sep 2025 06:22:17 -0700
Subject: [PATCH 6/9] apparmor: Fix double free of ns_name in
 aa_replace_profiles()

if ns_name is NULL after
1071         error = aa_unpack(udata, &lh, &ns_name);

and if ent->ns_name contains an ns_name in
1089                 } else if (ent->ns_name) {

then ns_name is assigned the ent->ns_name
1095                         ns_name = ent->ns_name;

however ent->ns_name is freed at
1262                 aa_load_ent_free(ent);

and then again when freeing ns_name at
1270         kfree(ns_name);

Fix this by NULLing out ent->ns_name after it is transferred to ns_name

Reported-by: Qualys Security Advisory <qsa@qualys.com>
Fixes: 04dc715e24d08 ("apparmor: audit policy ns specified in policy load")
Signed-off-by: John Johansen <john.johansen@canonical.com>
---
 security/apparmor/policy.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/security/apparmor/policy.c b/security/apparmor/policy.c
index cf1310f566b8..16cd51ae635f 100644
--- a/security/apparmor/policy.c
+++ b/security/apparmor/policy.c
@@ -1165,6 +1165,7 @@ ssize_t aa_replace_profiles(struct aa_ns *policy_ns, struct aa_label *label,
 				goto fail;
 			}
 			ns_name = ent->ns_name;
+			ent->ns_name = NULL;
 		} else
 			count++;
 	}
-- 
2.51.0

